zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -76,6 +79,160 @@ impl<F: Field> SameContextGadget<F> {
     }
 }
 
+/// Construction of step state transition that restores caller's state.
+#[derive(Clone, Debug)]
+pub(crate) struct RestoreContextGadget<F> {
+    caller_id: Cell<F>,
+    caller_is_root: Cell<F>,
+    caller_is_create: Cell<F>,
+    caller_code_hash: Cell<F>,
+    caller_program_counter: Cell<F>,
+    caller_stack_pointer: Cell<F>,
+    caller_gas_left: Cell<F>,
+    caller_memory_word_size: Cell<F>,
+    caller_reversible_write_counter: Cell<F>,
+}
+
+impl<F: Field> RestoreContextGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        rw_counter_delta: Expression<F>,
+        return_data_offset: Expression<F>,
+        return_data_length: Expression<F>,
+    ) -> Self {
+        // Read caller's context for restore
+        let caller_id = cb.call_context(None, CallContextFieldTag::CallerId);
+        let [caller_is_root, caller_is_create, caller_code_hash, caller_program_counter, caller_stack_pointer, caller_gas_left, caller_memory_word_size, caller_reversible_write_counter] =
+            [
+                CallContextFieldTag::IsRoot,
+                CallContextFieldTag::IsCreate,
+                CallContextFieldTag::CodeHash,
+                CallContextFieldTag::ProgramCounter,
+                CallContextFieldTag::StackPointer,
+                CallContextFieldTag::GasLeft,
+                CallContextFieldTag::MemorySize,
+                CallContextFieldTag::ReversibleWriteCounter,
+            ]
+            .map(|field_tag| cb.call_context(Some(caller_id.expr()), field_tag));
+
+        // Update caller's last callee information
+        for (field_tag, value) in [
+            (
+                CallContextFieldTag::LastCalleeId,
+                cb.curr.state.call_id.expr(),
+            ),
+            (
+                CallContextFieldTag::LastCalleeReturnDataOffset,
+                return_data_offset,
+            ),
+            (
+                CallContextFieldTag::LastCalleeReturnDataLength,
+                return_data_length,
+            ),
+        ] {
+            cb.call_context_lookup(true.expr(), Some(caller_id.expr()), field_tag, value);
+        }
+
+        // Consume all gas_left if call halts in exception
+        let gas_left = if cb.execution_state().halts_in_exception() {
```

Comment:
does line L137 represents logic in https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L238?  so for any non revert exception(e.g. OutOfGas), failed call consumes all gas left .

Reply:
Yes